core/src/main/java/org/mule/runtime/core/api/config/ConfigurationBuilderRegistryFilter.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2023 Salesforce, Inc. All rights reserved.
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.core.api.config;
+
+import org.mule.runtime.api.artifact.Registry;
+import org.mule.runtime.core.api.MuleContext;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+/**
+ * Configures the elements of the {@link Registry} of a {@link MuleContext}.
+ *
+ * @since 4.6.0
+ */
+public interface ConfigurationBuilderRegistryFilter {
+
+  /**
+   * Filters every object added to the target {@link MuleContext}'s registry, allowing customization when needed.
+   *
+   * @param serviceId           the ID of the object being added to the registry.
+   * @param serviceImplSupplier a {@link Supplier} for the object being added to the registry.
+   * @return the object to add to the registry, or {@link Optional#empty()} if no implementation should be added for the given
+   *         {@code serviceId}.
+   */
+  Optional<Object> filterRegisterObject(String serviceId, Supplier<Object> serviceImplSupplier);
```

Comment:
shouldn't this be called filtered object instead? Also, shouldn't it return Optional<Supplier<Object>> instead? There's no point in feeding it with a supplier otherwise

Reply:
You mean the method to be called something like `filteredObject` or the javadoc?
The idea of receiving a supplier is to be able to prevent the creation of objects that we won't use, or defer it. For example, in the `mule-framework` we make the Expression Manager lazy and this way we can create it only when it's needed. However, we need to put an actual object in the registry, so returning it right away seemed appropriate.